File: src/process-message.ts
import { Request, Response } from "express";
import express from "express";
import bodyParser from "body-parser";
import { Datastore } from "@google-cloud/datastore";
import { CloudTasksClient } from "@google-cloud/tasks";
import { environment } from "./environment";
import { getAudioTranscription } from "./openai";
import {
  getMediaMetadata,
  retrieveMedia,
  sendWhatsappMessage,
} from "./whatsapp";
import { AxiosError } from "axios";
import { PassThrough } from "stream";

const tasksClient = new CloudTasksClient();

const datastore = new Datastore();

type WSBaseMessage = {
  from: string;
  id: string;
  timestamp: string;
};

type WSTextMessage = {
  text: {
    body: string;
  };
  type: "text";
} & WSBaseMessage;

type WSAudioMessage = {
  audio: {
    id: string;
    mime_type: string;
  };
  type: "audio";
} & WSBaseMessage;

type WSMessage = WSTextMessage | WSAudioMessage;

type WSContact = { profile: { name: string }; wa_id: string };

type WSChange = {
  value?: {
    metadata?: {
      display_phone_number: string;
      phone_number_id: string;
    };
    contacts?: Array<WSContact>;
    messages?: Array<WSMessage>;
  };
  field: string;
};

type HookBody = {
  object?: string;
  entry?: Array<{
    id: string;
    changes?: Array<WSChange>;
  }>;
};

function handleAuthentication(req: Request, res: Response) {
  if (
    req.query["hub.mode"] !== "subscribe" ||
    req.query["hub.verify_token"] !== environment().WEBHOOK_VERIFY_TOKEN
  ) {
    res.sendStatus(400);
    return;
  }
  res.send(req.query["hub.challenge"]);
}

async function defineContactId(contactId: string, contacts: Array<WSContact>) {
  const contactKey = datastore.key(["Contact", contactId]);
  const [existingContact] = await datastore.get(contactKey);
  if (existingContact) return contactKey;

  const contact = contacts.find((contact) => contact.wa_id === contactId);

  const contactName = contact ? contact.profile.name : "Anonimo";
  const contactEntity = {
    key: contactKey,
    data: {
      name: contactName,
    },
  };
  await datastore.save(contactEntity);

  return contactKey;
}

async function appendToHistoryChat(
  message: WSBaseMessage,
  messageText: string,
  contacts: Array<WSContact>
) {
  const contactKey = await defineContactId(message.from, contacts);

  // Guarda el mensaje en Datastore
  const messageKey = datastore.key(["Message", message.id]);
  const messageEntity = {
    key: messageKey,
    data: [
      {
        name: "messageId",
        value: message.id,
      },
      {
        name: "from",
        value: contactKey,
      },
      {
        name: "text",
        value: messageText,
        excludeFromIndexes: true,
      },
      {
        name: "timestamp",
        value: Date.now(),
      },
      {
        name: "actor",
        value: "user",
      },
    ],
  };
  await datastore.save(messageEntity);
}

async function sendToQueue(contactId: string, messageId: string) {
  const queueName = environment().GC_TASK_CHAT_QUEUE;
  const location = environment().GC_TASK_LOCATION;
  const projectId = environment().GC_PROJECT_ID;
  const queuePath = tasksClient.queuePath(projectId, location, queueName);

  const task = {
    httpRequest: {
      httpMethod: "POST" as const,
      url: environment().WEBHOOK_QUEUE_PROCESSOR_URL,
      headers: {
        "Content-Type": "application/json",
      },
      body: Buffer.from(JSON.stringify({ contactId, messageId })).toString(
        "base64"
      ),
      scheduleTime: {
        seconds: Date.now() / 1000 + 60,
      },
      oidcToken: {
        serviceAccountEmail: environment().GC_SERVICE_ACCOUNT_EMAIL,
      },
    },
  };
  await tasksClient.createTask({ parent: queuePath, task });
}

async function getMessageText(message: WSMessage) {
  if (message.type === "text") return message.text.body;
  if (message.type === "audio") {
    const { stream } = await retrieveMedia(message.audio.id);
    const response = await getAudioTranscription(stream);
    const text = response.data.text;
    await sendWhatsappMessage(
      message.from,
      `Esto es lo que entendí en tu mensaje:\n\n*${text}*\n\nPor favor, dame un momento mientras reflexiono sobre la respuesta adecuada.`
    );
    return text;
  }
  return null;
}

async function processMessage(change: WSChange) {
  const message = change.value?.messages?.[0];
  const contacts = change.value?.contacts || [];
  if (!message) return;

  try {
    const messageText = await getMessageText(message);

    if (!messageText) {
      await sendWhatsappMessage(
        message.from,
        "Lo siento, no puedo entender este tipo de mensajes"
      );
      return;
    }

    await appendToHistoryChat(message, messageText, contacts);
    await sendToQueue(message.from, message.id);
  } catch (e) {
    if (e instanceof AxiosError) {
      console.error(e.response?.status);
      console.error(e.response?.statusText);
      console.error(JSON.stringify(e.response?.data));
    } else {
      console.error(e);
    }
    await sendWhatsappMessage(
      message.from,
      "¡Ups! Algo no está bien 🤒. Por favor, contacta al soporte técnico para que puedan resolver la situación lo más pronto posible."
    );
  }
}

async function handlePostRequest(req: Request, res: Response) {
  const body = req.body as HookBody;
  const changes = body.entry?.[0].changes || [];
  const changesWithMessages = changes.filter(
    (change) => !!change.value?.messages
  );

  if (changesWithMessages.length === 0) return res.sendStatus(200);

  await processMessage(changesWithMessages[0]);

  res.sendStatus(200);
}
const app = express();
app.use(bodyParser.json());

app.get("/message", handleAuthentication);
app.post("/message", handlePostRequest);
export default app;

File: src/process-message.py
from flask import Flask, request, jsonify
from google.cloud import datastore, tasks_v2
from google.protobuf import duration_pb2, timestamp_pb2
import os
import json

from environment import environment
from openai import get_audio_transcription
from whatsapp import mark_message_as_read, retrieve_media, send_whatsapp_message, notify_invalid_contact

app = Flask(__name__)
datastore_client = datastore.Client()
tasks_client = tasks_v2.CloudTasksClient()


def handle_authentication(req):
    if (
        req.args.get("hub.mode") != "subscribe"
        or req.args.get("hub.verify_token") != environment().get("WEBHOOK_VERIFY_TOKEN")
    ):
        return "", 400
    return jsonify(req.args.get("hub.challenge")), 200


async def define_contact_id(contact_id, contacts):
    contact_key = datastore_client.key("Contact", contact_id)
    existing_contact = datastore_client.get(contact_key)
    if existing_contact:
        return contact_key

    contact = next((contact for contact in contacts if contact["wa_id"] == contact_id), None)

    contact_name = contact["profile"]["name"] if contact else "Anonimo"
    contact_entity = datastore.Entity(key=contact_key)
    contact_entity.update({"name": contact_name})
    datastore_client.put(contact_entity)

    return contact_key


async def append_to_history_chat(message, message_text, contacts):
    contact_key = await define_contact_id(message["from"], contacts)

    message_key = datastore_client.key("Message", message["id"])
    message_entity = datastore.Entity(key=message_key)
    message_entity.update(
        {
            "messageId": message["id"],
            "from": contact_key,
            "text": message_text,
            "timestamp": datastore_client.transaction(),
            "actor": "user",
        }
    )
    datastore_client.put(message_entity)


async def send_to_queue(contact_id, message_id):
    queue_name = environment().get("GC_TASK_CHAT_QUEUE")
    location = environment().get("GC_TASK_LOCATION")
    project_id = os.getenv("PROJECT_ID", "GC_PROJECT_ID")
    queue_path = tasks_client.queue_path(project_id, location, queue_name)

    request_body = json.dumps({"contactId": contact_id, "messageId": message_id})
    task = {
        "http_request": {
            "http_method": "POST",
            "url": environment().get("WEBHOOK_QUEUE_PROCESSOR_URL"),
            "headers": {"Content-Type": "application/json"},
            "body": request_body.encode("utf-8").decode("latin-1"),
        },
        "schedule_time": (timestamp_pb2.Timestamp(seconds=int((time.time() + 10) * 1e9))),
        "oidc_token": {"service_account_email": environment().get("GC_SERVICE_ACCOUNT_EMAIL")},
    }
    tasks_client.create_task({"parent": queue_path, "task": task})


async def get_message_text(message):
    if message["type"] == "text":
        return message["text"]["body"]
    elif message["type"] == "audio":
        stream = await retrieve_media(message["audio"]["id"])
        response = await get_audio_transcription(stream)
        text = response["data"]["text"]
        await send_whatsapp_message(
            message["from"],
            f"Esto es lo que entendí en tu mensaje:\n*{text}*\nPor favor, dame un momento mientras reflexiono sobre la respuesta adecuada.",
        )
        return text
    else:
        return None


async def process_message(change):
    message = change["value"].get("messages", [])[0]
    contacts = change["value"].get("contacts", [])

    if not message:
        return

    if message["from"] not in environment().get("CONTACTS_WHITE_LIST", []):
        notify_invalid_contact(message["from"])
        return

    try:
        mark_message_as_read(message["id"])
        message_text = await get_message_text(message)

        if message_text is None:
            await send_whatsapp_message(message["from"], "Lo siento, no puedo entender este tipo de mensajes")
            return

        await append_to_history_chat(message, message_text, contacts)
        await send_to_queue(message["from"], message["id"])
    except Exception as e:
        print(e)
        await send_whatsapp_message(
            message["from"],
            "¡Ups! Algo no está bien 🤒. Por favor, contacta al soporte técnico para que puedan resolver la situación lo más pronto posible.",
        )


@app.route("/message", methods=["GET"])
def handle_get_request():
    return handle_authentication(request)


@app.route("/message", methods=["POST"])
async def handle_post_request():
    body = request.get_json(force=True)
    entry = body.get("entry", [{}])[0]

    if entry.get("id") != environment().get("WHATSAPP_ACCOUNT_ID"):
        return "", 200

    changes = entry.get("changes", [])
    changes_with_messages = [change for change in changes if "messages" in change.get("value", {})]

    if not changes_with_messages:
        return "", 200

    await process_message(changes_with_messages[0])

    return "", 200


if __name__ == "__main__":
    app.run()
